### PR TITLE
Service locator UTs

### PR DESCRIPTION
**Pull request by portellaa**

### CODE REVIEW

File: AlicerceTests/ServiceLocatorTests.swift [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    
+    var serviceLocator = ServiceLocator()
+    
+    override func setUp() {
+        super.setUp()
+        
+        serviceLocator = ServiceLocator()
+    }
+    
+    func testRegisterUsingServiceTypeInference() {
```
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: yes, makes a lot of sense. that is the problem when we use testing frameworks 😄 

👍

File: AlicerceTests/ServiceLocatorTests.swift [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+            let rDoubleService: Double = try serviceLocator.get()
+            let rIntService: Int = try serviceLocator.get()
+            
+            // Validate registered services
+            XCTAssertEqual(rStringService, testServiceString)
+            XCTAssertEqual(rDoubleService, testServiceDouble)
+            XCTAssertEqual(rIntService, testServiceInt)
+            
+        } catch let error {
+            XCTAssert(false, "💥: Register failed with error: \(error)")
```
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: thanks 😅

**Comment by p4checo** [comment]
We can guide ourselves using the following checklist for our `ServiceLocator` UTs, considering the current public API's:

- `register` (normal)
	+ Success
		* name ✅ 
		* type ✅ 
	+ Failure
		* duplicate service
		* duplicate lazy service

- `register` (lazy)
	+ Success
		* name ✅ 
		* type ✅ 
	+ Failure
		* duplicate service
		* duplicate lazy service

- `unregister`
	+ Success
		* name ✅ 
		* type ✅ 
	+ Failure
		* inexistent service
		* service type mismatch
		* lazy service type mismatch (WIP)

- `unregisterAll`
	+ Success ✅ 

- `get`
	+ Success
		* name
		* type
	+ Failure
		* inexistent service
		* service type mismatch
		* lazy service type mismatch (WIP)

After these I think we can consider trickier tests like recursiveness and other wacky use cases we can think of 😆

File: AlicerceTests/ServiceLocatorTests.swift [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+            XCTAssertEqual(rStringService, testLazyServiceString())
+            
+        } catch let error {
+            XCTAssert(false, "💥: Register failed with error: \(error)")
+        }
+    }
+    
+    func testRegisterUsingServiceTypeInference_GetUsingName() {
+        
+        typealias DoubleService = Double
```
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: fixed 👍

File: AlicerceTests/ServiceLocatorTests.swift [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+            XCTAssertEqual(rStringService, testServiceString)
+            XCTAssertEqual(rDoubleService, testServiceDouble)
+            
+            
+        } catch let error {
+            XCTAssert(false, "💥: Register failed with error: \(error)")
+        }
+    }
+    
+    func testRegisterUsingName_GetUsingTypeInference() {
```
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: sure, i can remove them but if we return the name when we register, we can do things like this

File: AlicerceTests/ServiceLocatorTests.swift [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            
+            // Check if it contains
+            XCTAssertEqual(rStringService, testServiceString)
+            
+            try serviceLocator.unregister(testServiceString)
+            
+            let _: String = try serviceLocator.get()
+            
+        } catch ServiceLocatorError.inexistentService {
+            XCTAssert(true, "🎉 Service removed with success")
```
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: wut?

File: AlicerceTests/ServiceLocatorTests.swift [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    
+    func testUnregisterAll() {
+
+        let testServiceString = "Test Register Service"
+        let testServiceDouble = 1.0
+        let testLazyServiceInt: () -> (Int) = {
+            return 1
+        }
+
+        func assert(with error: Error) {
```
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: 🥇

File: AlicerceTests/ServiceLocatorTests.swift (hunk 4 above)
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: Yes, like we discussed that brings some value to these `register` API's and also makes testing way easier 😉

File: AlicerceTests/ServiceLocatorTests.swift [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+            } catch ServiceLocatorError.inexistentService {
+                XCTAssert(true, "🎉 Service \(String(describing: type)) removed with success")
+            } catch let error {
+                assert(with: error)
+            }
+        }
+
+        // Test that services were all removed
+
+        checkServiceRemoved(Int.self)
```
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: Already discussed this in person. 💪 

OK, i agree with one dict for  the services but not the api for `list()` i think we don't need it.
Regarding the `enum`, we discussed we don't need the service `Type` for validation.

**Comment by portellaa** [comment]
@p4checo ready for review.

**Comment by portellaa** [comment]
I think i implemented all the tests, but will wait for your PR

File: AlicerceTests/ServiceLocatorTests.swift (hunk 7 above)
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: Agreed 😉👍

File: AlicerceTests/ServiceLocatorTests.swift [hunk 8]
```diff
@@ -65,12 +68,103 @@ class ServiceLocatorTests : XCTestCase {
...
+        do {
+
+            // Register the first service
+            try serviceLocator.register(service: testService)
+
+            // Register the same service, should fail with duplicateService
+            try serviceLocator.register(service: testService)
+
+        } catch ServiceLocatorError.duplicateService(_) {
+            print("✅"); return
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: Here we could test the returned error itself and check if the associated value (the conflicting service name) matches `serviceName`.

On another point: do we really need to print something here? The test will pass if we simply leave this `catch` block empty 😉 

On a more 🚓 mode: dat `;` and think we don't need the `return` 😜 

Finally, we have a lot of these cases on the test suite, so the same comment applies to all

File: AlicerceTests/ServiceLocatorTests.swift [hunk 9]
```diff
@@ -65,12 +68,103 @@ class ServiceLocatorTests : XCTestCase {
...
+        let testServiceLazy: () -> (String) = {
+            return "Test Register Service"
+        }
+
+        do {
+
+            // Register the first service (which should be lazy)
+            try serviceLocator.register(testServiceLazy)
+
+                // Register the a normal service, should fail with duplicateLazyService
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: bad alignment 🚓

File: AlicerceTests/ServiceLocatorTests.swift [hunk 10]
```diff
@@ -116,13 +210,110 @@ class ServiceLocatorTests : XCTestCase {
...
+
+        do {
+
+            // Register the first lazy service
+            try serviceLocator.register(testLazyServiceString)
+
+            // Force serviceLocator to call the lazy closure
+            let _: String = try serviceLocator.get()
+
+            // Register a early service, should fail with duplicateService
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: What's an "early service" ? 🤔 😆

File: AlicerceTests/ServiceLocatorTests.swift [hunk 11]
```diff
@@ -116,13 +210,110 @@ class ServiceLocatorTests : XCTestCase {
...
+            try serviceLocator.register(name: serviceName, testLazyServiceString)
+
+        } catch ServiceLocatorError.duplicateLazyServiceInit(_) {
+            print("✅")
+        } catch {
+            XCTFail("💥: Lazy Register failed with error: \(error)")
+        }
+    }
+
+    func testLazyRegisterWithNormal_UsingTypeInference_ShouldFailWithDuplicateService() {
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: This name is a bit confusing to me, especially the `"WithNormal" part.

If I understand correctly, this case tests whether a service registered as lazy using type inference that then gets initialised (and thus "updated" to normal ) should fail with `duplicateService` (and not `duplicateLazyService`) when another register attempt is made with the same name.

If this is correct, why not call these tests something like:

`testLazyRegister_UsingTypeInferenceAndInitialisingIt_ShouldFailWithDuplicateService`

What do you say?

File: AlicerceTests/ServiceLocatorTests.swift [hunk 12]
```diff
@@ -116,13 +210,110 @@ class ServiceLocatorTests : XCTestCase {
...
+            try serviceLocator.register(service: testService)
+
+        } catch ServiceLocatorError.duplicateService(_) {
+            print("✅")
+        } catch {
+            XCTFail("💥: Lazy Register failed with error: \(error)")
+        }
+    }
+
+    func testLazyRegisterWithNormal_UsingName_ShouldFailWithDuplicateService() {
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: Same as before

File: AlicerceTests/ServiceLocatorTests.swift [hunk 13]
```diff
@@ -135,8 +326,8 @@ class ServiceLocatorTests : XCTestCase {
             
             
             // Get registered services
-            let rStringService: String = try serviceLocator.get()
-            let rDoubleService: Double = try serviceLocator.get()
+            let rStringService: String = try serviceLocator.get(name: "String")
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: I don't agree with these `get`'s receiving a name, because the services aren't registered using a name and the caller *shouldn't* know what name is auto generated by the SL.

File: AlicerceTests/ServiceLocatorTests.swift [hunk 14]
```diff
@@ -189,11 +380,88 @@ class ServiceLocatorTests : XCTestCase {
...
         }
     }
+
+    // MARK: Failure
+
+    func testGet_UsingTypeInference_ShouldFailWithInexistingService() {
+
+        do {
+
+            // Check if there is a service with the same type
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: This comment seems incorrect

File: AlicerceTests/ServiceLocatorTests.swift [hunk 15]
```diff
@@ -189,11 +380,88 @@ class ServiceLocatorTests : XCTestCase {
...
+        }
+    }
+
+    func testGet_UsingName_ShouldFailWithInexistingService() {
+
+        let serviceName = "TestServiceName"
+
+        do {
+
+            // Check if there is a service with the same type
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: Same as above

File: AlicerceTests/ServiceLocatorTests.swift [hunk 16]
```diff
@@ -189,11 +380,88 @@ class ServiceLocatorTests : XCTestCase {
...
+        let serviceName = "Double"
+
+        do {
+
+            // Register a String type service
+            try serviceLocator.register(name: serviceName, service: testService)
+
+            // Get a Double type service
+            let _: Double = try serviceLocator.get()
+
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: extra newline 🚓

File: AlicerceTests/ServiceLocatorTests.swift [hunk 17]
```diff
@@ -189,11 +380,88 @@ class ServiceLocatorTests : XCTestCase {
...
+        let serviceName = "TestServiceName"
+
+        do {
+
+            // Register a String type service
+            try serviceLocator.register(name: serviceName, service: testService)
+
+            // Get a Double type service
+            let _: Double = try serviceLocator.get(name: serviceName)
+
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: same as above 🚓

File: AlicerceTests/ServiceLocatorTests.swift (hunk 10 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: 👦 🔎

File: AlicerceTests/ServiceLocatorTests.swift (hunk 8 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: 👏 🥇 ✅

File: AlicerceTests/ServiceLocatorTests.swift (hunk 9 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: 👮

File: AlicerceTests/ServiceLocatorTests.swift (hunk 11 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: 😂 poor guy 😄

File: AlicerceTests/ServiceLocatorTests.swift (hunk 13 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: Updated to use new api 🐭 💯

File: AlicerceTests/ServiceLocatorTests.swift (hunk 14 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: PO 👜 😄

**Comment by portellaa** [comment]
hey 🚓 addicted aka @p4checo, check my the last commit

**Comment by portellaa** [comment]
👮‍♀️  now it's ready 😝

File: AlicerceTests/ServiceLocatorTests.swift [hunk 18]
```diff
@@ -189,11 +383,129 @@ class ServiceLocatorTests : XCTestCase {
...
+        } catch ServiceLocatorError.serviceTypeMismatch(expected: let expectedType, found: let foundType) {
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: This could be a single `catch let (...)` 💄

File: AlicerceTests/ServiceLocatorTests.swift [hunk 19]
```diff
@@ -189,11 +383,129 @@ class ServiceLocatorTests : XCTestCase {
...
+            XCTAssertTrue(type(of: foundType) != type(of: expectedType))
```
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: As discussed, using `type(of:)` on an `X.Type` will give `X.Type.Type`, aka "typeception" 😆 

I've made some tests, and this seems to work:
```swift

(...)
        } catch let ServiceLocatorError.serviceTypeMismatch(expected: expectedType, found: foundType) {
            XCTAssert(expectedType != foundType)
            XCTAssert(expectedType == Double.self)
            XCTAssert(foundType == type(of: testService))
        }
(...)
```
Apparently `XCTAssertTrue` doesn't work on type comparisons (as well as `XCTAssertNotEqual` and `XCTAssertEqual`) 🤷‍♂️. Type comparison must use some special equality operators, I guess. 

I also added some more checks on the types themselves 😉

File: AlicerceTests/ServiceLocatorTests.swift (hunk 16 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: hey 👮‍♀️ this is fixed 😝

File: AlicerceTests/ServiceLocatorTests.swift (hunk 17 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: hey 👮‍♀️ this is fixed 😝

File: AlicerceTests/ServiceLocatorTests.swift (hunk 18 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: yeah i know, but i don't have a formalised idea which to use. but i will change, i also prefer the `let` before.

File: AlicerceTests/ServiceLocatorTests.swift (hunk 19 above)
Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: hum, ok i was using `true` or `equal` hum strange.

Comment (portellaa) [inline AlicerceTests/ServiceLocatorTests.swift]: regarding the test with double, i'm not sure if i agree with this one. We could besides that validate that the type is not the same as the `testService`

**Comment by portellaa** [comment]
New commit. Check the comments and see if you agree with this.

```
catch let ServiceLocatorError.serviceTypeMismatch(expected: expectedType, found: foundType) {
      let typeOfTestService = type(of: testService)
      XCTAssert(expectedType != foundType)
      XCTAssert(expectedType != typeOfTestService)
      XCTAssert(foundType == typeOfTestService)
}
```

File: AlicerceTests/ServiceLocatorTests.swift (hunk 19 above)
Comment (p4checo) [inline AlicerceTests/ServiceLocatorTests.swift]: The equality test on `expectedType` (`Double` in this case) is to test if the error type itself is well built, so I think it adds value.

By only testing against `testService`'s type, we aren't testing this.

Besides, the test `expectedType != type(of: testService)` can already be logically implied from the other two tests:

`expectedType != foundType` **and**  `expectedType == type(of: testService)` **implies** `foundType != type(of: testService)` 🤓 

Don't you agree?

**Comment by portellaa** [comment]
@p4checo i can't see your comment here 🤔 

but given that `testService` is a `String`, compare that the `expectedType` is not a `String` type but a different type. like a `Double` for example.

Both tests are valid, one is more explicit that the other. I was doing this when i got called to play  ⚽️ 🔨  😝 

```swift
private func service<T>(service: T, assertExpectedType expected: Any.Type, found: Any.Type) {
    XCTAssert(expected != found)
    XCTAssertTrue((expected is T) == false)
}
```

**Comment by portellaa** [comment]
@p4checo squashed

**Comment by p4checo** [comment]
Great work! 😄🍻
